examples/dashboard/dashboard.py

```diff
@@ -107,21 +116,68 @@ def downsample(aggregate, factor):
+                self.model.hover_tool.tooltips = tooltips
+
+            else:
+                hover_agg = downsample(agg.values, self.model.hover_size)
+                self.model.hover_source.data['value'] = hover_agg.flatten()
+                self.model.hover_tool.tooltips = [(self.model.field_title, '@value')]
+
+            # update legend ---------------------------------------------------------
+            if self.model.field in self.model.categorical_fields:
+
+                cat_dim = agg.dims[-1]
+                len_bar=900
+                cats = agg[cat_dim].values.tolist()
+                total = agg.sum(dim=cat_dim)
+                min_val = int(total.min().data)
+                max_val = int(total.max().data)
+                scale = np.linspace(min_val, max_val, 180, dtype=total.dtype)
+                cats = agg.coords[agg.dims[-1]].values
+                ncats = len(cats)
+                data = np.zeros((180, ncats, ncats), dtype=total.dtype)
+                data[:, np.arange(ncats), np.arange(ncats)] = scale[:, None]
+                cbar = DataArray(data, dims=['value', 'fake', cat_dim], 
+                                 coords=[scale, cats, cats])
+                img = tf.colorize(cbar, self.model.colormap, how=self.model.transfer_function)
+
+                dw = max_val - min_val
+                legend_fig = self.model.create_legend(img.values.T,
+                                                      x=min_val,
+                                                      y=0,
+                                                      dh=18 * ncats,
+                                                      dw=dw,
+                                                      x_start=min_val,
+                                                      x_end=max_val,
+                                                      y_range=(0,18 * ncats))
+                self.model.legend_vbox.children = [legend_fig]
+
+            else:
+                min_val = np.nanmin(agg.values)
+                max_val = np.nanmax(agg.values)
+                vals = np.linspace(min_val, max_val, 180)[None, :]
+                vals_arr = DataArray(vals)
+                img = tf.interpolate(vals_arr, cmap=self.model.color_ramp,
+                                     how=self.model.transfer_function)
+                dw = max_val - min_val
+                legend_fig = self.model.create_legend(img.values,
+                                                      x=min_val,
+                                                      y=0,
+                                                      dh=18,
+                                                      dw=dw,
+                                                      x_start=min_val,
+                                                      x_end=max_val,
+                                                      y_range=(0,18))
+
+                self.model.legend_vbox.children = [legend_fig]
```

Comment:
Once it settles down a bit, it would be good to move the legend support into a function or a class, with appropriate parameters, with an eye to eventually moving it out of the dashboard.py file and into a datashader library file.